BoxDynError in worker fn

Hi,
I'm  using boxed error in worker function.
<details>
<summary>example</summary>

```rust
use apalis::prelude::*;

use apalis_sql::sqlite::SqliteStorage;
use sqlx::SqlitePool;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    std::env::set_var("RUST_LOG", "debug,sqlx::query=info");
    tracing_subscriber::fmt::init();

    let pool = SqlitePool::connect("sqlite::memory:").await?;
    // Do migrations: Mainly for "sqlite::memory:"
    SqliteStorage::setup(&pool)
        .await
        .expect("unable to run migrations for sqlite");

    let storage: SqliteStorage<String> = SqliteStorage::new(pool.clone());

    Monitor::new()
        .register({
            WorkerBuilder::new("tasty-mango")
                .backend(storage)
                .build_fn(do_job)
        })
        .run()
        .await?;
    Ok(())
}

async fn do_job(job: String) -> Result<(), apalis::prelude::BoxDynError> {
    tracing::info!("job: {job}");
    Ok(())
}

#[derive(Debug)]
enum CustomError {}
impl std::error::Error for CustomError {}
impl std::fmt::Display for CustomError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        todo!()
    }
}
```
</details>

But i'm getting errors
<details>
<summary>errors</summary>

```
error[E0277]: the trait bound `ServiceFn<fn(std::string::String) -> impl Future<Output = Result<(), Box<(dyn StdError + Send + Sync + 'static)>>> {try_notify}, std::string::String, SqlContext, _>: tower_service::Service<apalis::prelude::Request<std::string::String, SqlContext>>` is not satisfied
  --> examples/sqlite/src/main.rs:25:18
   |
25 |                 .build_fn(try_notify)
   |                  ^^^^^^^^ unsatisfied trait bound
   |
   = help: the trait `tower_service::Service<apalis::prelude::Request<std::string::String, SqlContext>>` is not implemented for `ServiceFn<fn(String) -> ... {try_notify}, ..., ..., ...>`
   = help: the following other types implement trait `tower_service::Service<Request>`:
             ServiceFn<T, Req, Ctx, ()>
             ServiceFn<T, Req, Ctx, (A1, A2)>
             ServiceFn<T, Req, Ctx, (A1, A2, A3)>
             ServiceFn<T, Req, Ctx, (A1, A2, A3, A4)>
             ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5)>
             ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5, A6)>
             ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5, A6, A7)>
             ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5, A6, A7, A8)>
           and 9 others
   = note: required for `WorkerBuilder<String, SqlContext, SqliteStorage<String>, Identity, ...>` to implement `apalis::prelude::WorkerFactory<std::string::String, SqlContext, ServiceFn<fn(std::string::String) -> impl Future<Output = Result<(), Box<(dyn StdError + Send + Sync + 'static)>>> {try_notify}, std::string::String, SqlContext, _>>`
   = note: the full name for the type has been written to '/tmp/apalis/target/debug/deps/sqlite_example-de103ae0c3a1493e.long-type-8084250147342458497.txt'
   = note: consider using `--verbose` to print the full type name to the console
```
</details>

Replacing  boxed error with concrete enum fixes errors
```rust
async fn do_job(job: String) -> Result<(), CustomError> {
    tracing::info!("job: {job}");
    Ok(())
}

#[derive(Debug)]
enum CustomError {}
impl std::error::Error for CustomError {}
impl std::fmt::Display for CustomError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        todo!()
    }
}
```

apalis v0.6.0

Is there a reason for using BoxDynError?

I recommend using the concrete Error type provided in apalis like in email
service example or define your error type.

I will investigate a little more.


On Fri, 29 Nov 2024 at 11:50, boingrach ***@***.***> wrote:

> Hi,
> I'm using boxed error in worker function.
> example
>
> use apalis::prelude::*;
> use apalis_sql::sqlite::SqliteStorage;use sqlx::SqlitePool;
> #[tokio::main]async fn main() -> anyhow::Result<()> {
>     std::env::set_var("RUST_LOG", "debug,sqlx::query=info");
>     tracing_subscriber::fmt::init();
>
>     let pool = SqlitePool::connect("sqlite::memory:").await?;
>     // Do migrations: Mainly for "sqlite::memory:"
>     SqliteStorage::setup(&pool)
>         .await
>         .expect("unable to run migrations for sqlite");
>
>     let storage: SqliteStorage<String> = SqliteStorage::new(pool.clone());
>
>     Monitor::new()
>         .register({
>             WorkerBuilder::new("tasty-mango")
>                 .backend(storage)
>                 .build_fn(do_job)
>         })
>         .run()
>         .await?;
>     Ok(())}
> async fn do_job(job: String) -> Result<(), apalis::prelude::BoxDynError> {
>     tracing::info!("job: {job}");
>     Ok(())}
> #[derive(Debug)]enum CustomError {}impl std::error::Error for CustomError {}impl std::fmt::Display for CustomError {
>     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
>         todo!()
>     }}
>
> But i'm getting errors
> errors
>
> error[E0277]: the trait bound `ServiceFn<fn(std::string::String) -> impl Future<Output = Result<(), Box<(dyn StdError + Send + Sync + 'static)>>> {try_notify}, std::string::String, SqlContext, _>: tower_service::Service<apalis::prelude::Request<std::string::String, SqlContext>>` is not satisfied
>   --> examples/sqlite/src/main.rs:25:18
>    |
> 25 |                 .build_fn(try_notify)
>    |                  ^^^^^^^^ unsatisfied trait bound
>    |
>    = help: the trait `tower_service::Service<apalis::prelude::Request<std::string::String, SqlContext>>` is not implemented for `ServiceFn<fn(String) -> ... {try_notify}, ..., ..., ...>`
>    = help: the following other types implement trait `tower_service::Service<Request>`:
>              ServiceFn<T, Req, Ctx, ()>
>              ServiceFn<T, Req, Ctx, (A1, A2)>
>              ServiceFn<T, Req, Ctx, (A1, A2, A3)>
>              ServiceFn<T, Req, Ctx, (A1, A2, A3, A4)>
>              ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5)>
>              ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5, A6)>
>              ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5, A6, A7)>
>              ServiceFn<T, Req, Ctx, (A1, A2, A3, A4, A5, A6, A7, A8)>
>            and 9 others
>    = note: required for `WorkerBuilder<String, SqlContext, SqliteStorage<String>, Identity, ...>` to implement `apalis::prelude::WorkerFactory<std::string::String, SqlContext, ServiceFn<fn(std::string::String) -> impl Future<Output = Result<(), Box<(dyn StdError + Send + Sync + 'static)>>> {try_notify}, std::string::String, SqlContext, _>>`
>    = note: the full name for the type has been written to '/tmp/apalis/target/debug/deps/sqlite_example-de103ae0c3a1493e.long-type-8084250147342458497.txt'
>    = note: consider using `--verbose` to print the full type name to the console
>
> Replacing boxed error with concrete enum fixes errors
>
> async fn do_job(job: String) -> Result<(), CustomError> {
>     tracing::info!("job: {job}");
>     Ok(())}
> #[derive(Debug)]enum CustomError {}impl std::error::Error for CustomError {}impl std::fmt::Display for CustomError {
>     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
>         todo!()
>     }}
>
> apalis v0.6.0
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/geofmureithi/apalis/issues/470>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWXVRGQC5REFC4IPPE6I3RL2DATGRAVCNFSM6AAAAABSWUHUZGVHI2DSMVQWIX3LMV43ASLTON2WKOZSG4YDINBRGA3DQNI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

kinda
ErrorHandlerLayer works only for manual Service implementation
<detail>
<summary>example</summary>

```rust
use apalis::{layers::ErrorHandlingLayer, prelude::*};

use apalis_sql::{context::SqlContext, sqlite::SqliteStorage};
use futures::future::BoxFuture;
use sqlx::SqlitePool;
use tower::Service;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    tracing_subscriber::fmt().init();

    let pool = SqlitePool::connect("sqlite::memory:").await?;
    SqliteStorage::setup(&pool)
        .await
        .expect("unable to run migrations for sqlite");

    let storage: SqliteStorage<String> = SqliteStorage::new(pool.clone());

    let service_worker = WorkerBuilder::new("do_job")
        .layer(ErrorHandlingLayer::new())
        .backend(storage.clone())
        // fixed with layer
        .build(ServiceWithBoxedError);

    Monitor::new().register(service_worker).run().await?;

    let fn_worker = WorkerBuilder::new("do_job")
        .layer(ErrorHandlingLayer::new())
        .backend(storage.clone())
        // errors
        .build_fn(do_job);

    Monitor::new().register(fn_worker).run().await?;

    Ok(())
}

async fn do_job(job: String) -> Result<(), Box<dyn std::error::Error + Sync + Send + 'static>> {
    tracing::info!("job: {job}");
    Ok(())
}

struct ServiceWithBoxedError;

impl Service<Request<String, SqlContext>> for ServiceWithBoxedError {
    type Response = ();
    type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(
        &mut self,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Result<(), Self::Error>> {
        todo!()
    }

    fn call(&mut self, req: Request<String, SqlContext>) -> Self::Future {
        todo!()
    }
}

```
</detail>

the fn case fails with errors
<detail>
<summary>errors</summary>

```
error[E0277]: the trait bound `apalis::prelude::ServiceFn<fn(std::string::String) -> impl futures::Future<Output = Result<(), Box<(dyn std::error::Error + std::marker::Send + Sync + 'static)>>> {do_job}, std::string::String, SqlContext, _>: Service<apalis::prelude::Request<std::string::String, SqlContext>>` is not satisfied
```
</detail>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,10 +12,7 @@ All notable changes to this project are documented in this file.
 - **deps** : update rust crate redis to 0.31 ([#555](https://github.com/geofmureithi/apalis/pull/555))
 - **deps** : update rust crate sentry-core to 0.38.0 ([#569](https://github.com/geofmureithi/apalis/pull/569))
 - **deps** : update rust crate criterion to 0.6.0 ([#574](https://github.com/geofmureithi/apalis/pull/574))
-
-
-## Fixed
-
+- **error-handling** : ease the error type that is returned by a worker function ([#577](https://github.com/geofmureithi/apalis/pull/577))
 - **PostgresStorage**: fix type error when updating jobs ([#539](https://github.com/geofmureithi/apalis/issues/539))
 
 ## [0.7.1](https://github.com/geofmureithi/apalis/releases/tag/v0.7.1)

--- a/examples/basics/src/main.rs
+++ b/examples/basics/src/main.rs
@@ -30,18 +30,6 @@ async fn produce_jobs(storage: &SqliteStorage<Email>) {
 }
 
 #[derive(thiserror::Error, Debug)]
-pub enum ServiceError {
-    #[error("data store disconnected")]
-    Disconnect(#[from] std::io::Error),
-    #[error("the data for key `{0}` is not available")]
-    Redaction(String),
-    #[error("invalid header (expected {expected:?}, found {found:?})")]
-    InvalidHeader { expected: String, found: String },
-    #[error("unknown data store error")]
-    Unknown,
-}
-
-#[derive(thiserror::Error, Debug)]
 pub enum PanicError {
     #[error("{0}")]
     Panic(String),
@@ -54,7 +42,7 @@ async fn send_email(
     svc: Data<EmailService>,
     worker: Worker<Context>,
     cache: Data<ValidEmailCache>,
-) -> Result<(), ServiceError> {
+) -> Result<(), BoxDynError> {
     info!("Job started in worker {:?}", worker.id());
     let cache_clone = cache.clone();
     let email_to = email.to.clone();

--- a/packages/apalis-core/src/response.rs
+++ b/packages/apalis-core/src/response.rs
@@ -1,7 +1,7 @@
-use std::{any::Any, fmt::Debug, sync::Arc};
+use std::{fmt::Debug, sync::Arc};
 
 use crate::{
-    error::Error,
+    error::{BoxDynError, Error},
     task::{attempt::Attempt, task_id::TaskId},
 };
 
@@ -133,21 +133,17 @@ impl IntoResponse for bool {
     }
 }
 
-impl<T: Any, E: std::error::Error + Sync + Send + 'static + Any> IntoResponse
-    for std::result::Result<T, E>
-{
+impl<T, E: Into<BoxDynError>> IntoResponse for std::result::Result<T, E> {
     type Result = Result<T, Error>;
     fn into_response(self) -> Result<T, Error> {
         match self {
             Ok(value) => Ok(value),
             Err(e) => {
-                // Try to downcast the error to see if it is already of type `Error`
-                if let Some(custom_error) =
-                    (&e as &(dyn std::error::Error + 'static)).downcast_ref::<Error>()
-                {
+                let e = e.into();
+                if let Some(custom_error) = e.downcast_ref::<Error>() {
                     return Err(custom_error.clone());
                 }
-                Err(Error::Failed(Arc::new(Box::new(e))))
+                Err(Error::Failed(Arc::new(e)))
             }
         }
     }

